tests/test_visualization/test_hierarchical_heatmap.py

```diff
@@ -1,98 +1,126 @@
+            'feature3': [3, 5, 7, 9, 11, 13, 15, 17]
         })
 
         annotation = pd.DataFrame({
             'phenotype': [
                 'phenotype1',
                 'phenotype1',
                 'phenotype2',
-                'phenotype2'
+                'phenotype2',
+                'phenotype3',
+                'phenotype3',
+                'phenotype4',
+                'phenotype4'
             ]
         })
 
         self.adata = anndata.AnnData(X=X, obs=annotation)
 
-    def test_returns_correct_types(self):
-        """Test if correct types are returned."""
-        mean_intensity, matrixplot = hierarchical_heatmap(
-            self.adata,
-            'phenotype'
-        )
-        self.assertIsInstance(mean_intensity, pd.DataFrame)
-        self.assertIsInstance(matrixplot, MatrixPlot)
-
-    def test_invalid_annotation_raises_error(self):
-        """Test if invalid annotation raises error."""
-        err_msg = (r"The annotation 'invalid_annotation' does not exist "
-                   r"in the provided AnnData object. Available annotations "
-                   r"are: \['phenotype'\]")
-        with self.assertRaisesRegex(KeyError, err_msg):
-            hierarchical_heatmap(self.adata, 'invalid_annotation')
-
     def test_nan_annotation_raises_error(self):
         """Test if NaN annotation raises error."""
         self.adata.obs['phenotype'] = [
             None,
             'phenotype1',
             'phenotype2',
-            'phenotype1'
+            'phenotype1',
+            'phenotype3',
+            'phenotype3',
+            'phenotype4',
+            'phenotype4'
         ]
         with self.assertRaises(ValueError):
             hierarchical_heatmap(self.adata, 'phenotype')
 
-    def test_invalid_layer_raises_error(self):
-        """Test if invalid layer raises error."""
-        err_msg = ("The layer 'invalid_layer' does not exist in the "
-                   "provided AnnData object")
-        with self.assertRaisesRegex(KeyError, err_msg):
-            hierarchical_heatmap(
-                self.adata,
-                'phenotype',
-                layer='invalid_layer'
-            )
-
     def test_calculates_correct_mean_intensity(self):
         """Test if correct mean intensities are calculated."""
-        expected_mean_intensity = self.adata.to_df().groupby(
-            self.adata.obs['phenotype']
-        ).mean().reset_index()
-        mean_intensity, _ = hierarchical_heatmap(self.adata, 'phenotype')
-        self.assertEqual(mean_intensity.shape, (2, 3))
+
+        # Ensure annotation column is categorical
+        mean_intensity, _, _ = hierarchical_heatmap(self.adata, 'phenotype')
+        self.assertTrue(
+            pd.api.types.is_categorical_dtype(self.adata.obs['phenotype'])
+        )
+
+        # Calculate the expected mean intensity
+        intensities = self.adata.to_df()
+        labels = self.adata.obs['phenotype']
+        expected_mean_intensity = (
```

Comment:
@fangliu117, in the unit test, please hardcode the ground truth values. Otherwise, the unit test will be running the exact same function as the implementation.

Reply:
Yes, I have refactored the code with hardcoded expected mean intensities.